src/ir/ExpressionManipulator.cpp

```diff
@@ -28,7 +30,7 @@ Expression* flexibleCopy(Expression* original, Module& wasm, CustomCopier custom
 
     Builder builder;
 
-    Copier(Module& wasm, CustomCopier custom) : wasm(wasm), custom(custom), builder(wasm) {}
+    Copier(Module& wasm, CustomCopier custom) : wasm(wasm), custom(std::move(custom)), builder(wasm) {}
```

Comment:
why is it doing a move of a std::function?

Reply:
std::function is still an object, and it can be moved. Since std::function can be arbitrarily complex/expensive to copy, it is better to move it.

https://stackoverflow.com/a/13680662/1953079

Reply:
Why is it valid to move it, though? In theory a user could pass in a std::function and be surprised by the fact that it is no longer valid afterwards.

If the overhead is an issue, we can pass it in by reference, `std::function<..>&`?

Reply:
Because the input argument `custom` is passed by value, and will only live the lifetime on the constructor anyway.   (Took me a while to see the pattern here but I think it makes sense).

Reply:
I see. Still isn't `&` clearer? We can use it on the external interface as well, to avoid any passing by value (if we think that's high-overhead - is it actually, though?)

Reply:
When you pass by const ref, you always do a copy of an object.

When you do pass by value, then move, then for rvalue objects you can do double move, which is more preferable (faster) for most types (because when you copy, you need to allocate a memory, when you move - you don't) https://stackoverflow.com/a/37443572/1953079

Reply:
I don't follow. In the example there,
```
Box::Box(const Box & other)
{
   // copy the contents of other
}
```
There is a const ref, and we used that to refer to `other`. We then create a new Box, copying those fields. We don't hang on to `other`, we just look at it to initialize the fields of the new Box.

In our case here, though, we are not in a constructor like that - we just want to pass around a  `std::function` from place to place. We could pass around a raw pointer to it, in other words, avoiding any copying at all. I believe that is functionally what happens if all the locations have `std::function<..>&`.